plugin/procfs/procfs.cc

```diff
@@ -0,0 +1,475 @@
+/*
+(C) 2019 Percona LLC and/or its affiliates
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation; version 2 of the License.
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
+*/
+
+/**
+ @file
+ ProcFS plugin is intended to provide access to Linux performance counters
+ by running SQL queries. This is useful for Cloud and MySQL-as-service
+ installations.
+ The plugin reads files from /proc and /sys filesystems and returns
+ contents as is. All files not specified in datadir/procfs.cnf are ignored.
+ You can write patterns in glob(7) style line by line:
+ /sys/block/sd[a-z]/stat
+ /proc/version*
+
+ Clients could fetch individual files:
+ SELECT * FROM INFORMATION_SCHEMA.PROCFS WHERE FILE = '/proc/version';
+
+ To install this plugin, copy the .so file to the plugin directory and do
+ INSTALL PLUGIN procfs SONAME 'procfs.so';
+*/
+
+#include <m_ctype.h> /* my_charset_bin */
+#include <mysql_version.h>
+#if MYSQL_VERSION_ID >= 80002
+#include <sql/sql_class.h> /* THD, Security context */
+#include <sql/item_cmpfunc.h>
+#include <mf_wcomp.h>
+bool schema_table_store_record(THD *thd, TABLE *table);
+#else
+#include <sql_class.h> /* THD, Security context */
+#include <item_cmpfunc.h>
+#include <sql_show.h> /* schema_table_store_record */
+#endif
+#include <mysql/psi/mysql_thread.h>
+
+#include <stdlib.h>
+#include <dirent.h>
+#include <fnmatch.h>
+#include <time.h>
+#include <ctype.h>
+#include <mysql/plugin.h>
+
+#include <fstream>
+#define IS_PROCFS_CONTENTS_SIZE 60000
+#define IS_PROCFS_REFRESH_SEC 60
+#define IS_PROCFS_FILES_PER_PATTERN 10000
+#define IS_PROCFS_PATTERN_DEPTH 10
+
+namespace procfs_plugin {
+std::vector<std::string> files;
```

Comment:
Using an `std::shared_ptr<std::vector<std::string>>` instead, and creating new vectors instead of clearing the existing vector during a file refresh would allow the fill method to aquire a read lock for filling only for a short time. This could increase the response time when multiple sessions use this table.

Another improvement after this could be moving the filelist refresh to a background thread, which would mean that the client threads only have to read the actual files. (assuming that globbing proc and sys is slow because of the huge amount of files)